orttraining/orttraining/core/graph/pattern_graph/pattern_graph.cc

```diff
@@ -0,0 +1,351 @@
+    res.Clear();
+    std::unordered_set<const Node*> graph_path, pattern_path;
+    std::unordered_map<const Node*, const Node*> path_map;
+    if (FindMatchRecursively(node, pattern_root, graph_path, pattern_path, path_map, res, target_graph)) {
+      std::unordered_map<std::string, MatchedNodeGroup> matches;
+      matches.emplace(pattern_root->Name(), MatchedNodeGroup(target_graph.GetNode(node_index), name_pnode_mapping_[pattern_root->Name()]));
+      res.AppendToNodeGroups(matches);
+      return Status::OK();
+    }
+  }
+
+  return Status(common::ONNXRUNTIME, common::FAIL, "No match for the target graph.");
+}
+
+Status PatternGraph::GetNodeDefs() {
+  for (size_t i = 0; i < pgraph_nodes_.size(); ++i) {
+    node_defs_.push_back(pgraph_nodes_[i].GetNodeDef());
+  }
+  for (size_t i = 0; i < pgraph_inputs_.size(); ++i) {
+    node_defs_.push_back(pgraph_inputs_[i].GetNodeDef());
+  }
+  return Status::OK();
+}
+
+Status PatternGraph::ToGraphInternal() {
+  ORT_ENFORCE(GetNodeDefs().IsOK());
+  /** Todo: p_initializer_names_to_preserve should contains all initializer names in the pattern graphs.
+   * to avoid constant folding by any opportunaties.
+   */
+  const std::unordered_set<std::string>* p_initializer_names_to_preserve = nullptr;
+  GraphAugmenter::GraphDefs graph_defs;
+  graph_defs.AddNodeDefs(node_defs_);
+  auto status = GraphAugmenter::AugmentGraph(ort_model_ptr_->MainGraph(), graph_defs, p_initializer_names_to_preserve);
+  ORT_ENFORCE(status.IsOK());
+  return Status::OK();
+}
+
+/**
+ * Try to find a match for all the args of a node. We just use brute force here to iterate all the possible cases.
+ * Let's assume that args of the node in pattern graph are a1, a2, ..., ai, ... an and that of target graph is b1, b2, ..., bj, ... bn
+ * Then we keep the order of {ai} fixed and try to match it with {bj}.
+ * We first choose b1, then choose an arg in {bj | j != 1, j <= n} as next arg and the recurse it.
+ * "visited" is used to record visited args to avoid duplicated conditions.
+ */
+bool FindMatchForArgs(const Graph& graph,
+                      PatternGraph& pattern_graph,
+                      const std::unordered_map<std::string, const PGraphInput*>& name_pargs_mapping,
+                      ConstPointerContainer<std::vector<NodeArg*>>& p_args,
+                      ConstPointerContainer<std::vector<NodeArg*>>& t_args,
+                      const std::unordered_map<std::string, std::unique_ptr<ArgCompareFunc>>& arg_constraints,
+                      const std::unique_ptr<ArgCompareFunc>& arg_default_constraint,
+                      size_t p_arg_idx,
+                      std::unordered_set<const NodeArg*>& visited) {
+  /* TODO: constraint the order of the args */
+  if (p_arg_idx >= p_args.size()) return true;
+  auto p_arg = p_args[p_arg_idx];
+  PARSER_LOG << "Try to find match for arg " << p_arg->Name();
+  auto p_arg_name = p_arg->Name();
+  auto find_defined_arg = name_pargs_mapping.find(p_arg_name);
```

Comment:
in which case we will go to this branch,

"  auto p_arg_name = p_arg->Name();
  auto find_defined_arg = name_pargs_mapping.find(p_arg_name);
  if (find_defined_arg == name_pargs_mapping.end())
    return FindMatchForArgs(graph, pattern_graph, name_pargs_mapping, p_args, t_args, arg_constraints, arg_default_constraint, p_arg_idx + 1, visited);"

Reply:
It's the case that it meets the output of a previous node. For example, if we have the graph below, the ```x``` and ```y``` are nodes and the ```p``` is an initializer. Though we can also match the output arg of ```x```, that's actually the responsibility of the "node match" part. Therefore, if we meet an arg name that is not in the list of the args the user define, we just skip it for the next match.

```
 x     p
  \   /
    y
```